core/chains/evm/logpoller/log_poller.go

```diff
@@ -276,163 +269,177 @@ func (lp *logPoller) backfill(ctx context.Context, start, end int64) int64 {
+		lp.lggr.Infow("Unfinalized log query", "logs", len(logs), "currentBlockNumber", currentBlockNumber, "blockHash", currentBlock.Hash())
+		err2 = lp.orm.q.WithOpts(pg.WithParentCtx(ctx)).Transaction(func(tx pg.Queryer) error {
+			if err3 := lp.orm.InsertBlock(currentBlock.Hash(), currentBlock.Number().Int64(), pg.WithQueryer(tx)); err3 != nil {
 				return err3
 			}
 			if len(logs) == 0 {
 				return nil
 			}
-			return lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs))
+			return lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs), pg.WithQueryer(tx))
 		})
 		if err2 != nil {
-			// If we're unable to insert, don't increment currentBlockNumber and just retry
-			lp.lggr.Warnw("Unable to save logs, retrying", "err", err2, "block", currentBlock.Number())
-			return currentBlockNumber
+			lp.lggr.Warnw("Unable to save logs resuming from last saved block + 1", "err", err2, "block", currentBlock.Number())
+			return
 		}
 		currentBlockNumber++
 	}
-	return currentBlockNumber
 }
 
-func (lp *logPoller) findLCA(h common.Hash) (int64, error) {
-	// Find the first place where our chain and their chain have the same block,
-	// that block number is the LCA.
-	block, err := lp.ec.BlockByHash(context.Background(), h)
+// Find the first place where our chain and their chain have the same block,
+// that block number is the LCA. Return the block after that, where we want to resume polling.
+func (lp *logPoller) findBlockAfterLCA(ctx context.Context, current *types.Block) (*types.Block, error) {
+	// Current is where the mismatch starts.
+	// Check its parent to see if its the same as ours saved.
+	parent, err := lp.ec.BlockByHash(ctx, current.ParentHash())
 	if err != nil {
-		return 0, err
+		return nil, err
 	}
-	blockNumber := block.Number().Int64()
-	startBlockNumber := blockNumber
-	for blockNumber >= (startBlockNumber - lp.finalityDepth) {
-		ourBlockHash, err := lp.orm.SelectBlockByNumber(blockNumber)
+	blockAfterLCA := *current
+	reorgStart := parent.Number().Int64()
+	// We expected reorgs up to the block after (current - finalityDepth),
+	// since the block at (current - finalityDepth) is finalized.
+	// We loop via parent instead of current so current always holds the LCA+1.
+	// If the parent block number becomes < the first finalized block our reorg is too deep.
+	for parent.Number().Int64() >= (reorgStart - lp.finalityDepth) {
+		ourParentBlockHash, err := lp.orm.SelectBlockByNumber(parent.Number().Int64(), pg.WithParentCtx(ctx))
 		if err != nil {
-			return 0, err
+			return nil, err
 		}
-		if block.Hash() == ourBlockHash.BlockHash {
-			// If we do have the blockhash, that is the LCA
-			return blockNumber, nil
+		if parent.Hash() == ourParentBlockHash.BlockHash {
+			// If we do have the blockhash, return blockAfterLCA
+			return &blockAfterLCA, nil
 		}
-		blockNumber--
-		block, err = lp.ec.BlockByHash(context.Background(), block.ParentHash())
+		// Otherwise get a new parent and update blockAfterLCA.
```

Comment:
This is the main change here: we now loop backwards keeping two pointers "parent" and its child "blockAfterLCA". The parent is used to check for LCA detection, then once we find it its child will be blockAfterLCA which is where we want to resume polling from. Just simplifies the logic upstream since then we have the full block to resume polling from.